receiver/podmanreceiver/podman.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2022 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !windows
+// +build !windows
+
+package podmanreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/podmanreceiver"
+
+import (
+	"context"
+	"encoding/json"
+	"net/url"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+type clientFactory func(logger *zap.Logger, cfg *Config) (PodmanClient, error)
+
+type PodmanClient interface {
+	ping(context.Context) error
+	stats(context.Context, url.Values) ([]ContainerStats, error)
+	list(context.Context, url.Values) ([]Container, error)
+	events(context.Context, url.Values) (<-chan Event, <-chan error)
+}
+
+type ContainerScraper struct {
+	client         PodmanClient
+	containers     map[string]Container
+	containersLock sync.Mutex
+	logger         *zap.Logger
+	config         *Config
+}
+
+func NewContainerScraper(engineClient PodmanClient, logger *zap.Logger, config *Config) *ContainerScraper {
+	return &ContainerScraper{
+		client:     engineClient,
+		containers: make(map[string]Container),
+		logger:     logger,
+		config:     config,
+	}
+}
+
+// Containers provides a slice of Container to use for individual FetchContainerStats calls.
+func (pc *ContainerScraper) Containers() []Container {
+	pc.containersLock.Lock()
+	defer pc.containersLock.Unlock()
+	containers := make([]Container, 0, len(pc.containers))
+	for _, container := range pc.containers {
+		containers = append(containers, container)
+	}
+	return containers
+}
+
+// LoadContainerList will load the initial running container maps for
+// inspection and establishing which containers warrant stat gathering calls
+// by the receiver.
+func (pc *ContainerScraper) LoadContainerList(ctx context.Context) error {
+	params := url.Values{}
+	runningFilter := map[string][]string{
+		"status": {"running"},
+	}
+	jsonFilter, err := json.Marshal(runningFilter)
+	if err != nil {
+		return nil
+	}
+	params.Add("filters", string(jsonFilter[:]))
```

Comment:
Should `pc.client.list` take `filter map[string][]string` as the second argument in order to avoid duplicating this code here and in `InspectAndPersistContainer`

Reply:
Good catch! I definitely prefer to use a `map[string][]string` in those cases and make the `pc.client` responsible of generating the `url.Values{}` type

Reply:
My only concern would be in case we want to add any other query parameter in the request apart from `filters`. Indeed, you can also provide the `pod` and a `limit` in the query parameters. See: https://docs.podman.io/en/latest/_static/api.html#operation/ContainerListLibpod

In that case I would prefer to keep the client as flexible as possible and do not contain specific behavior (generating the "filters" parameter). If duplicated code becomes a problem I would opt for creating a helper function to generate a url parameter for a `map[string][]string`. What do you think?

Reply:
I'm ok with keeping all the client methods taking params for consistency